### ISSUE TITLE
Support for CryptoKey objects as private keys

### ISSUE THREAD

**Issue by vlad-ignatov** [smart-on-fhir/client-js#154]
A while back @jmandel wrote:

> Ideally we'd support non-extractable keys in the browser context, meaning we'd provide a way to pass a CryptoKey object into the library as a config param (in addition to the current option of passing a raw extracted private JWK)

This seems strait-forward but introduces several challenges:
1. IE support may be impossible (but supposedly is not needed)
2. Persistence - currently all options are passed to `authorize`. Then they are serialized and persisted in `SessionStorage` so that they can be used after the redirect. Non-extractable keys cannot be stored this way. The recommended way to store them is in `IndexedDB`, which complicates things. Another option would be to change the API and to allow `ready` to also accept parameters. This should be possible because the private key is not needed in the initial stage of the auth flow, thus there is no reason to pass it to `authorize`. This also implies that the developer will be responsible for obtaining (and persisting) the key object, and the library will simply use it, if provided.
3. The `alg` property is optional in JWK but required for the SMART Asymmetric Auth flow. Given a `CryptoKey` object, we can probably infer what its `alg` should be, but I have the feeling that is not reliable.
4. The `kid` property cannot be determined from a `CryptoKey` object without extracting, but it is also required in our case. Not sure how to proceed with this one.

**Comment by jmandel** [comment]
1. I don't see any reason to support asymmetric client authentication in IE.

2. It's not just "recommended" - my understanding is that IndeedDb is the only technically valid way to persist non extractable keys. But we could handle this case in the ways you suggest (allowing args to ready, or storing this particular property in IndexedDb).
3. Good call. It's okay to require a CryptoKey and an additional `alg` parameter if we need that to disambiguate.
4. As in (3) -- could even combine these into a `publicJwk` argument, which would have the full public key, alg, and kid...

**Comment by vlad-ignatov** [comment]
So, we could start with something like this:
```ts
interface ReadyOptions {
    privateKey?: JWK | {
        key: CryptoKey
        alg: "RS384" | "ES384"
        kid: string
    };
    clientPublicKeySetUrl?: string; // for jku
    [key: string]: any; // Other options TBD
}

FHIR.oauth2.ready({
    privateKey: {
        key: CryptoKeyObject,
        alg: "RS384",
        kid: "1234"
    },
    clientPublicKeySetUrl: "https://whatever"
}).then(client => ... )
```
Then, depending on if and how people use that, we may also consider `IndexedDb`. 

My only question so far is why did you name this "publicJwk argument" above, instead of private - by mistake, or am I missing something?

**Comment by jmandel** [comment]
>So, we could start with something like this

Yes, that would be a safe way to start

>My only question so far is why did you name this "publicJwk argument" above, instead of private - by mistake, or am I missing something?

Oh, I was suggesting something like

```ts
FHIR.oauth2.ready({
    privateKey:  CryptoKeyObject,
    publicJwk: {
      "kty": "EC",
      "kid": "3Kfdg-XwP-7gXyywtUfUADwBumDOPKMQx-iELL11W9s",
      "use": "sig",
      "alg": "ES256",
      "crv": "P-256",
      "x": "11XvRWy1I2S0EyJlyf_bWfw_TQ5CJJNLw78bHXNxcgw",
      "y": "eZXwxvO1hvCY0KucrPfKo7yAyMT6Ajc3N7OkAB6VYy8",
    },
    clientPublicKeySetUrl: "https://whatever"
```
Just because the algorithm and the KID are already defined in that data model, but I am not pushing strongly for this suggestion.

**Comment by vlad-ignatov** [comment]
Looks like we are talking about the same thing. In my example, if `privateKey` has a `key` property, then it should have the shape `{ key: CryptoKey, alg: "RS384" | "ES384", kid: string }`. Otherwise, `privateKey` is expected to be valid `JWK` having `alg`, `kid` and everything else.

What confuses me is that you have both `privateKey` and `publicJwk`. What is the idea behind using these together and why is that jwk public?

**Comment by jmandel** [comment]
Your suggestion works well! (I hadn't fully understood it when I suggested separately supplying the alg/kid in the context of a public key input.)

**Comment by vlad-ignatov** [comment]
This is now released in `v2.5.0`
